crates/subspace-farmer/src/bin/subspace-farmer/commands/cluster/controller.rs

```diff
@@ -242,39 +286,56 @@ pub(super) async fn controller(
                 .await
             }
         },
-        "controller-farms".to_string(),
+        "farms".to_string(),
     )?;
 
-    let caches_fut = run_future_in_dedicated_thread(
-        move || async move {
-            maintain_caches(
-                &cache_group,
-                &nats_client,
-                farmer_cache,
-                CACHE_IDENTIFICATION_BROADCAST_INTERVAL,
-            )
-            .await
-        },
-        "controller-caches".to_string(),
-    )?;
+    let caches_fut = farmer_caches
+        .iter()
+        .cloned()
+        .zip(cache_groups)
+        .enumerate()
+        .map(|(index, (farmer_cache, cache_group))| {
+            let nats_client = nats_client.clone();
+
+            async move {
+                let fut = run_future_in_dedicated_thread(
+                    move || async move {
+                        maintain_caches(
+                            &cache_group,
+                            &nats_client,
+                            &farmer_cache,
+                            CACHE_IDENTIFICATION_BROADCAST_INTERVAL,
+                        )
+                        .await
+                    },
+                    format!("caches-{index}"),
+                );
+                anyhow::Ok(fut?.await?)
+            }
+        })
+        .collect::<FuturesUnordered<_>>();
 
     let networking_fut = run_future_in_dedicated_thread(
         move || async move { node_runner.run().await },
-        "controller-networking".to_string(),
+        "networking".to_string(),
     )?;
 
     Ok(Box::pin(async move {
         // This defines order in which things are dropped
         let networking_fut = networking_fut;
         let farms_fut = farms_fut;
-        let caches_fut = caches_fut;
-        let farmer_cache_worker_fut = farmer_cache_worker_fut;
+        let mut caches_fut = caches_fut;
+        let caches_fut = caches_fut.next().map(|result| result.unwrap_or(Ok(Ok(()))));
+        let mut farmer_cache_workers_fut = farmer_cache_workers_fut;
+        let farmer_cache_workers_fut = farmer_cache_workers_fut
+            .next()
+            .map(|result| result.unwrap_or(Ok(())));
```

Comment:
`Result::unwrap_or(Ok(_))` ignores outer errors by replacing them with `Ok(_)`.
Did you want `Result::flatten()` or its equivalent `Result::and_then(identity)` instead?

Reply:
I don't think it ignores any errors. It actually unwraps `None`, which can only happen when there are no cache groups configured, which in turn is not possible because while CLI option is `Vec<String>`, it has the default value, so for practical purposes can't be empty. This is essentially a placeholder that will not really happen unless there is something that I didn't foresee, in which case it'll exit with unexpected cache exit message.